lib/chef/resource/windows_ad_join.rb

```diff
@@ -169,6 +169,20 @@ def on_desired_domain?
           node_domain == new_resource.domain_name.downcase
         end
 
+        #
+        # @return [String] the correct user and domain to use.
+        #   if the domain_user property contains an @ symbol followed by any number of non white space characheters
+        #   then we assume it is a user from another domain than the one specifed in the resource domain_name property.
+        #   if this is the case we do not append the domain_name property to the domain_user property
+        #   regex: https://rubular.com/r/VzupyBWemC8MMM
+        def sanitize_usename
+          if new_resource.domain_user =~ /^\S*@\S*$/
```

Comment:
I don't know much about AD names, but the use of `*` here seems a bit fishy to me.  This maches:

`foo@`
`@`
`@foo`

But not:

`foo bar@baz`

I think it would also be good to link to the rules we think the domain_user needs to comply to. Do they allow spaces? If so, this regex will be incorrect I think.  Given the current regex, I think we would be better off with checking if the string contains '@' rather than using a regex because then the behavior would be easier to document and understand -- assuming that is compatible with whatever naming rules apply here.

Reply:
Did some digging:
The UPN seems to be based off RFC 822 (which is now obsolete)
 https://docs.microsoft.com/en-us/windows/win32/ad/naming-properties#userprincipalname

https://tools.ietf.org/html/rfc822#section-6
is the address part of the rfc that UPN is based off.
However, It is not clear (to me) if white spaces are allowed by this spec
So I agree that matching on the @ is preferable to the current regex, given its simplicity in an unclear area.